Add document about Open Source requirements (#127)
operate-first/community: open-source.md

---
title: Minimum Open Source requirements for Services

> description: The minimum bar for Open Source when working on Services, that then allows further progression.

Reviewer: ```suggestion
description: The minimum requirements for Open Source when working on Services, that then allows further progression.
```
- `bar` is a more of an idiomatic phrase and less clear, more difficult to translate.
- I have come around to agreeing that although it looks awkwardly Germanic, we should be capitalizing Open Source as part of controlling the narrative around what Open Source means. I'm not sure though about `Services` being capitalized unless you also define what this special instance of the word services means.

Author: Indeed my German first language is creeping in here. I've removed capitalization from the word Service, except in headings where the heading is not a full sentence.

---

# Minimum Open Source Requirements for Services

> Here we document the minimum bar for use of Open Source in services. When a service meets this minimum bar, it is then possible for contributors to further progress the service toward the complete Operate First methodology. This represents what we expect from any service claiming to be Open Source.

Reviewer: Note it's a good practice to put a sentence alone on each line for easier to read diffs. This will be part of our docs standard for the project.

Reviewer: ```suggestion
Here we document the minimum requirements for use of Open Source in services. When a service meets this minimum requirement, it is then possible for contributors to further progress the service toward the complete Operate First methodology. This represents what we expect from any _service_ claiming to be Open Source.
```

Author: Split sentences ... and changed usage of word "bar" to "requirements" throughout.

## Requirements

These requirements do not represent best practices, but are the bare minimum for Open Source development of services. Most Open Source models (beyond “throw it over the wall” behavior) are covered by these requirements in one way or another:

 1. **All code and assets necessary to operate the service are shared under an open source license, and publicly accessible.**
 1. **A public contributor can use the same workflow as a typical team member to make a change to the service.**

For many projects the above seem obvious; this is how we generally work together on Open Source projects. Nevertheless such requirements are being explicitly specified, since services are a new area of work, and since the services and their sources are often not “distributed” for users to operate, and optionally modify, their own copy.

## Frequently Asked Questions:

> Due to the variety of services, projects, teams and workflows there is a lot of flexibility in the requirements. Secondly, the requirements are a bare minimum, and best practices usually go beyond the requirements. Thus we must address common questions about the requirements:

Reviewer: ```suggestion
Due to the variety of services, projects, teams, and workflows there is a lot of flexibility in the requirements. Secondly, the requirements are a bare minimum, and best practices usually go beyond the requirements. Thus we must address common questions about the requirements:
```
I'm a strict adherent of the serial comma (aka Oxford comma) and hope it becomes standard for this project, so I'm fixing them now in anticipation of that. :)

Author: Added comma.

### Q: What “assets” should be Open Source?

> **A:** Make everything open by default, except where law, security, privacy or common sense says otherwise. A general rule of thumb is if you commit it to a repository, it should be open. Assets that should not be shared, are ones where you don’t have redistribution rights (such as a RHEL VM image, or private information, personal identifiable information), or things that are actually secret: Don’t share keys, passwords or tokens.

Reviewer: ```suggestion
**A:** Make everything open by default, except where law, security, privacy, or common sense says otherwise. A general rule of thumb is if you commit it to a repository, it should be open. Assets that should not be shared are ones where you don’t have redistribution rights (such as a RHEL VM image, private information, and personal identifiable information), or things that are actually secret: Don’t share keys, passwords, or tokens.
```

Author: Added changes.

If changing configuration is an activity that a typical team member does in order to change the service, then you should share that configuration publicly, in order to meet **requirement #2**.

> It is *recommended* that you include sufficient configuration to operate the service, or some base default version of configuration. Putting secrets in your configuration is a bad practice that should be avoided, regardless of Open Source.

Author: Updated to use underscore for italics.

### Q: Should tests be shared with the code?

> **A:** Tests that are used during operation the service (as per **requirement #1**), and tests that used when run when making a change to the service (as per **requirement #2**), *must be shared publicly* under an Open Source license.

Reviewer: ```suggestion
**A:** Tests that are used during operation the service (as per **requirement #1**), and tests that used when run when making a change to the service (as per **requirement #2**), _must be shared publicly_ under an Open Source license.
```
The to-be-published markup standards differentiate italic emphasis with the underscores so it is easier to see the difference in the unrendered MarkDown file, e.g. in a plain text editor that does not show MarkDown styling inline while editing.

Author: Updated to use underscore for italics.

> It is *recommended* that tests which are independent of the service (eg: penetration tests or performance tests) are Open Source, but this is not required.

Author: Updated to use underscore for italics.

### Q: Should anyone be able to trigger CI or test suites?

> **A:** The **requirement #2** above does *not require* that any public contributor can trigger the test suite, if that activity is limited to certain known contributors.

Reviewer: ```suggestion
**A:** The **requirement #2** above does _not require_ that any public contributor can trigger the test suite, if that activity is limited to certain known contributors.
```

Author: Updated to use underscore for italics.

> It is common for Open Source projects to limit which contributors can trigger CI or run test suites on their provided infrastructure. This is usually done to prevent abuse such as exfiltrating secrets or denial of service. It’s *recommended* that your project has a mechanism whereby a known contributor, or maintainer, can review a contribution from an arbitrary contributor and then allow the CI or test suite to run on it.

Reviewer: ```suggestion
It is common for Open Source projects to limit which contributors can trigger CI or run test suites on their provided infrastructure. This is usually done to prevent abuse such as exfiltrating secrets or denial of service. It’s _recommended_ that your project has a mechanism whereby a known contributor, or maintainer, can review a contribution from an arbitrary contributor and then allow the CI or test suite to run on it.
```

Author: Updated to use underscore for italics.
